### PR TITLE
feat: automataic versioning

### PR DESCRIPTION
**Pull request by dbeal-eth**
fixes https://linear.app/usecannon/issue/CAN-500/automatic-versioning

NOTE: merging auto deployments into this branch because the changes ended up being related.

### CODE REVIEW

File: packages/builder/src/package.ts [hunk 1]
```diff
@@ -209,3 +211,49 @@ export async function publishPackage({
...
+  let oldDeployHash: string | null = null;
+  let oldDelpoyTimestamp = 0;
+
+  await Promise.all(
+    deployers.map(async (addr) => {
+      const [deployTimestamp, deployHash] = (
+        await storeRead(
+          runtime.provider,
+          addr,
+          viem.keccak256(viem.stringToBytes(`${packageReference.name}:${packageReference.preset}`))
```
Comment (mjlescano) [inline packages/builder/src/package.ts:231]: can it use the @ separator? just to be consistent with packageRef

File: packages/builder/src/utils/onchain-store.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
+/*
+contract Storage {
```
Comment (mjlescano) [inline packages/builder/src/utils/onchain-store.ts:2]: It should use a more descriptive name, like CannonDeploymentsHistory

File: packages/website/src/features/Deploy/QueueFromGitOpsPage.tsx [hunk 3]
```diff
@@ -807,8 +852,8 @@ function QueueFromGitOps() {
                     w="100%"
                     onClick={() => {
                       execTxn.writeContract(stager.executeTxnConfig!, {
-                        onSuccess: async () => {
-                          await router.push(links.DEPLOY);
+                        onSuccess: () => {
+                          router.push(links.DEPLOY);
```
Comment (mjlescano) [inline packages/website/src/features/Deploy/QueueFromGitOpsPage.tsx:856]: why this change?

File: packages/website/package.json [hunk 4]
```diff
@@ -48,8 +48,8 @@
     "@types/react": "18.2.37",
     "@types/react-dom": "18.2.6",
     "@usecannon/api": "workspace:*",
-    "@usecannon/builder": "2.17.0",
-    "@usecannon/cli": "2.17.0",
+    "@usecannon/builder": "workspace:*",
+    "@usecannon/cli": "workspace:*",
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply packages/website/package.json]: maybe... I was having issues with it using the wrong version of the package

**Review by mjlescano — APPROVED** [review]

File: packages/builder/src/schemas.ts [hunk 5]
```diff
@@ -892,6 +892,17 @@ export const chainDefinitionSchema = z
          * Keywords for search indexing
          */
         keywords: z.array(z.string()).describe('Keywords for search indexing'),
+        /**
+         * Any deployers that could publish this package. Will be used for automatic version management.
+         */
+        deployers: z
+          .array(
+            z.string().refine((val) => !!val.match(RegExp(/^0x[a-fA-F0-9]{40}$/, 'gm')), {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply packages/builder/src/schemas.ts:900]: yea i'm just copying and pasting what we have elsewhere so we end up with this. but I agree :+1:
